File: LC/python/Partition List.py
class Solution:
    # @param head, a ListNode
    # @param x, an integer
    # @return a ListNode
    def partition(self, head, x):
        if head == None or head.next == None:
            return head
        
        dummy = ListNode(0)
        tmp = dummy
        
        dummy1 = ListNode(0)
        less = dummy1
        
        p = head
        
        while p:
            if p.val < x:
                tmp.next = p
                tmp = tmp.next
                p = p.next
                tmp.next = None
            else:
                less.next = p
                less = less.next
                p = p.next
                less.next = None            
        
        tmp.next = dummy1.next
        return dummy.next

File: LC/java/Partition List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode dummy,tmp,larger,l;
        dummy = new ListNode(0);
        larger = new ListNode(0);
        l = larger;
        tmp = dummy;
        
        while(head!=null){
            if(head.val<x){
                tmp.next = head;
                head=head.next;
                tmp = tmp.next;
                tmp.next = null;
                
            }
            else{
                larger.next = head;
                head = head.next;
                larger = larger.next;
                larger.next = null;
            }
        }
        tmp.next = l.next;
    
        return dummy.next;
            
    }
}